bluesky_widgets/qt/threading.py

```diff
@@ -0,0 +1,625 @@
+        self._pause_interval = 0.01
+
+    def work(self) -> None:
+        """Core event loop that calls the original function.
+
+        Enters a continual loop, yielding and returning from the original
+        function.  Checks for various events (quit, pause, resume, etc...).
+        (To clarify: we are creating a rudimentary event loop here because
+        there IS NO Qt event loop running in the other thread to hook into)
+        """
+        while True:
+            if self.abort_requested:
+                self.aborted.emit()
+                break
+            if self._paused:
+                if self._resume_requested:
+                    self._paused = False
+                    self._resume_requested = False
+                    self.resumed.emit()
+                else:
+                    time.sleep(self._pause_interval)
+                    continue
+            elif self._pause_requested:
+                self._paused = True
+                self._pause_requested = False
+                self.paused.emit()
+                continue
+            try:
+                self.yielded.emit(self._gen.send(self._next_value()))
+            except StopIteration as exc:
+                return exc.value
+
+    def send(self, value: Any):
+        """Send a value into the function (if a generator was used)."""
+        self._incoming_value = value
+
+    def _next_value(self) -> Any:
+        out = None
+        if self._incoming_value is not None:
+            out = self._incoming_value
+            self._incoming_value = None
+        return out
+
+    @property
+    def is_paused(self) -> bool:
+        """Whether the worker is currently paused."""
+        return self._paused
+
+    def toggle_pause(self) -> None:
+        """Request to pause the worker if playing or resume if paused."""
+        if self.is_paused:
+            self._resume_requested = True
+        else:
+            self._pause_requested = True
+
+    def pause(self) -> None:
+        """Request to pause the worker."""
+        if not self.is_paused:
+            self._pause_requested = True
+
+    def resume(self) -> None:
+        """Send a request to resume the worker.
+        """
+        if self.is_paused:
+            self._resume_requested = True
+
+
+############################################################################
+
+# public API
+
+# For now, the next three functions simply wrap the QThreadPool API, and allow
+# us to track and cleanup all workers that were started with ``start_worker``,
+# provided that ``wait_for_workers_to_quit`` is called at shutdown.
+# In the future, this could wrap any API, or a pure python threadpool.
+
+
+def set_max_thread_count(num: int):
+    """Set the maximum number of threads used by the thread pool.
+
+    Note: The thread pool will always use at least 1 thread, even if
+    maxThreadCount limit is zero or negative.
+    """
+    QThreadPool.globalInstance().setMaxThreadCount(num)
```

Comment:
No issue here, Just a high level inquiry/concern: As a stand alone application exposing access to setting these variables is fine, but when it is underneath another larger application. Could this potential for issues? For example: Napari sets set_max_thread_count() then some developer sees this in bluesky_widgets and sets the same thing using a different function in a different location. Now there is a conflict. Not saying this will happen, also I am not entirely sure what a good solution looks like. But potentially exposing functions that has side-effects from dependencies will mess with the overall program execution. At least, using a global shared pool definitely has some pros versus everyone creating their own local threadpool and producing crazy overhead in thread switching.

Reply:
It's a good point. Ultimately getting this right will be up to the person who is managing the QApplication instance. I think the best we can do is leave clear code comments in examples where we use this.